### PR TITLE
[DCA][Staggered] init health check

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot/pull/2476

### CODE REVIEW

File: Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py [hunk 1]
```diff
@@ -251,6 +254,23 @@ def get_is_symbol_wildcard(cls) -> bool:
...
+    async def single_exchange_process_health_check(self, chained_orders: list, tickers: dict) -> list:
+        created_orders = []
+        if await self._should_rebalance_orders():
+            target_asset = exchange_util.get_common_traded_quote(self.exchange_manager)
+            created_orders += await self.single_exchange_process_optimize_initial_portfolio([], target_asset, tickers)
+            for producer in self.producers:
+                await producer.trigger_staggered_orders_creation()
+        return created_orders
+
+    async def _should_rebalance_orders(self):
```
Comment (Herklos) [inline Trading/Mode/staggered_orders_trading_mode/staggered_orders_trading.py:266]: :+1:

File: Trading/Mode/grid_trading_mode/grid_trading.py [hunk 2]
```diff
@@ -42,6 +42,7 @@ class GridTradingMode(staggered_orders_trading.StaggeredOrdersTradingMode):
     USER_COMMAND_PAUSE_ORDER_MIRRORING = "pause orders mirroring"
     USER_COMMAND_TRADING_PAIR = "trading pair"
     USER_COMMAND_PAUSE_TIME = "pause length in seconds"
+    SUPPORTS_HEALTH_CHECK = False   # WIP   # set True when self.health_check is implemented
```
Comment (Herklos) [inline Trading/Mode/grid_trading_mode/grid_trading.py:45]: Why isn't it the same implementation as staggered?

Comment (GuillaumeDSM) [reply Trading/Mode/grid_trading_mode/grid_trading.py:45]: staggered have min and max prices boundaries, it makes orders "grid" translation much more complicated

Comment (Herklos) [reply Trading/Mode/grid_trading_mode/grid_trading.py:45]: Oh right, I see

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 3]
```diff
@@ -706,10 +708,36 @@ def init_user_inputs(self, inputs: dict) -> None:
...
 
+        self.is_health_check_enabled = self.UI.user_input(
+            self.ENABLE_HEALTH_CHECK, commons_enums.UserInputTypes.BOOLEAN,
+            self.is_health_check_enabled, inputs,
+            title="Health check: when enabled, OctoBot will automatically sell traded assets that are not associated "
+                  "to a sell order and that represent at least the 'Health check threshold' part of the "
+                  "portfolio. Health check can be useful to avoid inactive funds, for example if a buy order got "
+                  "filled but no sell order was created. Requires a common quote market for each traded pair. "
+                  "Warning: will sell any asset associated to a trading pair that is not covered by a sell order, "
+                  "even if not bought by OctoBot or this trading mode.",
```
Comment (Herklos) [inline Trading/Mode/dca_trading_mode/dca_trading.py:724]: :+1:

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 4]
```diff
@@ -740,17 +768,34 @@ async def single_exchange_process_optimize_initial_portfolio(
...
+        if (
+            self.exchange_manager.is_backtesting
+            or common_quote is None
+            or not (self.use_take_profit_exit_orders or self.use_stop_loss)
+        ):
+            # skipped when:
+            # - backtesting
+            # - common_quote is unset
+            # - not using take profit or stop losses, health check should not be used
+            return []
```
Comment (Herklos) [inline Trading/Mode/dca_trading_mode/dca_trading.py:781]: Shouldn't we override `should_trigger_health_check`?

File: Trading/Mode/dca_trading_mode/tests/test_dca_trading_mode.py [hunk 5]
```diff
@@ -846,6 +846,115 @@ async def test_single_exchange_process_optimize_initial_portfolio(tools):
...
+        portfolio["ETH"] = trading_personal_data.SpotAsset("ETH", eth_holdings, eth_holdings)
+        orders = await mode.single_exchange_process_health_check([], {})
+        assert len(orders) == 1
+        sell_order = orders[0]
+        assert isinstance(sell_order, trading_personal_data.SellMarketOrder)
+        assert sell_order.symbol == "ETH/USDT"
+        assert sell_order.origin_quantity == eth_holdings
+        assert portfolio["ETH"].total == trading_constants.ZERO
+        after_eth_usdt_portfolio = portfolio["USDT"].total
+        assert after_eth_usdt_portfolio > after_btc_usdt_portfolio
```
Comment (Herklos) [inline Trading/Mode/dca_trading_mode/tests/test_dca_trading_mode.py:913]: :+1:

File: Trading/Mode/dca_trading_mode/dca_trading.py (hunk 4 above)
Comment (GuillaumeDSM) [reply Trading/Mode/dca_trading_mode/dca_trading.py:781]: should_trigger_health_check renamed in https://github.com/Drakkar-Software/OctoBot-Trading/pull/1006

**Review by Herklos — APPROVED** [review]

**Pull request merged**